API Adding default_classes config option for Form and FormField

dhensby opened the pull request:
This adds the ability to add default classes to `Form` and `FormField` which is useful for CSS frameworks to ensure fields and forms can have classes which are relevant.

Fixes #3688

stevie-mayhew commented on tests/forms/FormFieldTest.php:
```diff
@@ -3,7 +3,52 @@
+		$field = new FormField('MyField');
+
+		$this->assertContains('class1', $field->extraClass(), 'Default class not added to class list');
+
```
Shouldn't this be "default classes added to class list"?

dhensby replied:
The message in the final parameter is shown if the assertion fails.

So, in this example, we would state the message if the `extraClass` method didn't contain "class1" and therefore is a failure because the default class wasn't added.

stevie-mayhew replied:
I knew I was going mental. Thanks @dhensby

tractorcow replied:
Aren't they normally written in the positive sense? https://github.com/silverstripe/silverstripe-framework/blob/3.1/tests/model/DataObjectTest.php#L1205

dhensby replied:
Erm, It's going to be hard to write this failure positively. I could write "Class list does not contain expected default class"/

kinglozzer replied:
You’re writing what you expect to happen, so it’d just be: “Default class is added to class list”

dhensby replied:
The final parameter is the message that is shown when the assertion fails. In this example that would be when the expected class "class1" is no in the string returned by the function.

If the assertion failed, you're suggesting the message should be "Default class is added to class list", which is not why the test failed.

If no final argument is provided, then it would use the default which would be: "Failed to assert that string "[insert returned string]" contained "class1".

kinglozzer replied:
@dhensby I agree with you: the negative version makes more sense than the positive when displayed in the error output, but every other core test is written in the positive sense (I was just pointing out “our” convention for those messages).

We’re [not alone](https://github.com/symfony/DependencyInjection/blob/master/Tests/ContainerTest.php#L99) in messages being written this way, but the wording on [PHPUnit assertion docs](https://phpunit.de/manual/current/en/appendixes.assertions.html#appendixes.assertions.assertArrayHasKey) always starts:

> Reports an error identified by $message... 

Our messages don’t really identify the error, so perhaps they should all be rewritten the other way around?

dhensby replied:
Maybe we just shouldn't customise the messages ;)

tractorcow commented on forms/Form.php:
```diff
@@ -142,6 +142,11 @@ class Form extends RequestHandler {
 	protected $extraClasses = array();
 
 	/**
+	 * @var array $default_classes The default classes to apply to the Form
```
Can you pop a `@config` notifier here if it's configurable?

We haven't, but probably should add this to the coding conventions.

chillu commented on forms/Form.php:
```diff
@@ -264,6 +271,19 @@ public function setupFormErrors() {
+		$defaultClasses = self::config()->get('default_classes');
+		if ($defaultClasses) {
+			foreach ($defaultClasses as $class) {
+				$this->addExtraClass($class);
```
Indentation ;)

dhensby replied:
What's up with the indentation? :(

chillu commented:
Awesome feature, this will make theming a lot easier

dhensby commented:
Ok, I've addressed the comments by:

- Rewording the test assertion messages
- Adding `@config` to the static vars

@chillu if you can tell me what's up with the indentation, then I'll sort that out. Otherwise, I think we're ready.
